lib/spack/spack/schema/mirrors.py

```diff
@@ -10,19 +10,60 @@
 """
 from typing import Any, Dict
 
+
+string_or_variable = {
+    "oneOf": [
+        {
+            "type": "object",
+            "required": ["variable"],
+            "additionalProperties": False,
+            "properties": {"variable": {"type": "string"}},
+        },
+        {"type": "string"},
+    ]
+}
+
+
 #: Common properties for connection specification
 connection = {
     "url": {"type": "string"},
     # todo: replace this with named keys "username" / "password" or "id" / "secret"
     "access_pair": {
-        "type": "array",
-        "items": {"type": ["string", "null"], "minItems": 2, "maxItems": 2},
+        "oneOf": [
+            {
+                "type": "array",
+                "items": {"minItems": 2, "maxItems": 2, **string_or_variable},
+            },  # deprecated
+            {
+                "type": "object",
+                "required": ["secret_variable"],
+                # Only allow id or id_variable to be set, not both
+                "oneOf": [{"required": ["id"]}, {"required": ["id_variable"]}],
+                "properties": {
+                    "id": {"type": "string"},
+                    "id_variable": {"type": "string"},
+                    "secret_variable": {"type": "string"},
+                },
+            },
+        ]
     },
-    "access_token": {"type": ["string", "null"]},
     "profile": {"type": ["string", "null"]},
     "endpoint_url": {"type": ["string", "null"]},
+    "access_token": {"type": ["string", "null"]},  # deprecated
+    "access_token_variable": {"type": ["string", "null"]},
```

Comment:
Replace the `access_token` with `access_token_variable`. The alternate syntax here was too many newlines for a field that wasn't going to get any additional options. If we someday want to add `file:` here maybe worth revisiting, but I think the current consensus is we don't want to support secrets on the FS.